src/ipfs_retrieval.py

```
@@ -169,25 +201,32 @@ def get_metadata_by_key(key: str) -> dict:
+        # Fallback to local cache if endpoint is unreachable or JSON is malformed
+        cache_file = os.path.join(os.path.dirname(__file__), "cids.json")
+        if os.path.exists(cache_file):
+            try:
+                with open(cache_file, "r") as f:
+                    json_cid = json.load(f)  # can raise JSONDecodeError
+                return list(json_cid.keys())
+            except json.JSONDecodeError:
+                # local file is corrupt or empty
+                raise RuntimeError(
+                    "Failed to retrieve dataset list from endpoint or local cache."
+                )
```

Ruff (0.8.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)